habitat/datasets/rearrange/samplers.py

```diff
@@ -0,0 +1,618 @@
+        while (
+            targets_found < target_number
+            and num_pairing_tries < self.max_sample_attempts
+        ):
+            num_pairing_tries += 1
+            new_object = self.single_sample(sim, snap_down, vdb)
+            if new_object is not None:
+                targets_found += 1
+                found_match = False
+                for object_instance in self.object_instance_set:
+                    if (
+                        object_instance.creation_attributes.handle
+                        == new_object.creation_attributes.handle
+                        and object_instance.handle not in new_target_objects
+                    ):
+                        new_target_objects[object_instance.handle] = new_object
+                        found_match = True
+                        # remove this object instance match from future pairings
+                        self.object_set.remove(
+                            new_object.creation_attributes.handle
+                        )
+                        break
+                assert (
+                    found_match is True
+                ), "Failed to match instance to generated object. Shouldn't happen, must be a bug."
+
+        if len(new_target_objects) >= self.num_objects[0]:
+            return new_target_objects
+
+        # we didn't find all placements, so remove all new objects and return
+        logger.info(
+            f"Failed to sample all target placements in {self.max_sample_attempts} tries."
+        )
+        logger.info(
+            f"    Only able to sample {len(new_target_objects)} targets out of {len(self.object_instance_set)}..."
+        )
+        # cleanup
+        for new_object in new_target_objects.values():
+            sim.get_rigid_object_manager().remove_object_by_handle(
+                new_object.handle
+            )
+        return None
+
+
+class ArticulatedObjectStateSampler:
+    def __init__(
+        self, ao_handle: str, link_name: str, state_range: Tuple[float, float]
+    ) -> None:
+        self.ao_handle = ao_handle
+        self.link_name = link_name
+        self.state_range = state_range
+        assert self.state_range[1] >= self.state_range[0]
+
+    def sample(
+        self, sim: habitat_sim.Simulator
+    ) -> Optional[
+        Dict[habitat_sim.physics.ManagedArticulatedObject, Dict[int, float]]
+    ]:
+        """
+        For all matching AOs in the scene, sample and apply the joint state for this sampler.
+        Return a list of tuples (instance_handle, link_name, state)
+        """
+        ao_states: Dict[
+            habitat_sim.physics.ManagedArticulatedObject, Dict[int, float]
+        ] = {}
+        # TODO: handle sampled invalid states (e.g. fridge open into wall in some scenes)
+        aom = sim.get_articulated_object_manager()
+        # get all AOs in the scene with the configured handle as a substring
+        matching_ao_instances = aom.get_objects_by_handle_substring(
```

Comment:
Good point. This loop is intended to catch multiple instances (e.g. `fridge_0000` and `fridge_0001`) but it would also currently catch `fridge2_0000` and if that object also had a link with the same name it would be modified. We should additionally cull urdf file names which don't exactly match the input.